Add `--config-override` feature

sriniv27 opened the pull request:
<!--
Thank you for wanting to contribute!

Please fill out this description, and the checklist below.

Here are some key points to include in your description:
- What is this new code intended to do?
- Are there any related issues?
- What is the motivation for this change?
- What is an example of usage, or changes to config files? (if applicable)
-->
- What is this new code intended to do?
  - This submission is intended to provide a mechanism to override the stored configuration for the lifetime of the current invocation of `jrnl` by way of a  delimited list of key-value pairs. The keys can be provided in dot-notation form e.g. `colors.body` 
- Are there any related issues?
  #1157 
- What is the motivation for this change?
  I would like to run jrnl with different configurations for some specific workflows but still maintain my current configuration for "normal" writing workflows.
  For example:  
  - My default writing configuration is set so that I use `code -w` as my editor. For rapid logging, I want to be able open a terminal and be dropped into the stdin prompt, record my thought, and Ctrl+D out. 
  - In tmux have a separate, narrow, vertical pane that has the contents of my day's log, for which I might want a smaller `linewrap` value than the default
  - See discussion in #1157 for other examples/ideas
- What is an example of usage, or changes to config files? (if applicable)
   See [motivating comment](https://github.com/jrnl-org/jrnl/issues/1157#issuecomment-770296603)
  `jrnl -on today --config-override editor '' --config-override journals.log "~/logfile.txt"`
  `jrnl -on today --config-override linewrap 30 --config-override highlight false --config-override display_format fancy`
### Checklist

- [x] I have read the [contributing doc](https://github.com/jrnl-org/jrnl/blob/develop/CONTRIBUTING.md).
- [x] I have included a link to the relevant issue number.
- [x] I have tested this code locally.
- [x] I have checked to ensure there aren't other open [pull requests](../pulls)
  for the same issue.
- [x] I have written new tests for these changes, as needed.
- [x] All tests pass.
<!--
NOTE: Your PR may not be reviewed if there are any failing tests. You can run
tests locally with `make test` (see the contributing doc if you need help with
`make`), or use our automated tests after you submit your PR.
-->

sriniv27 committed:
add test and argument handler for runtime override of configurations.

sriniv27 committed:
identify location to apply override in "main"

sriniv27 committed:
update gitignore

sriniv27 committed:
remove unneeded import

sriniv27 committed:
add jrnl interface test for overriden configurations

sriniv27 committed:
trivial whitespace change

sriniv27 committed:
implement runtime override

sriniv27 committed:
make format

sriniv27 committed:
refactor override unittest

sriniv27 committed:
clean up unused import

sriniv27 committed:
start writing integration test

sriniv27 committed:
add linewrap override scenario

sriniv27 committed:
implement editor override step

sriniv27 committed:
add dev dependencies on pytest -mock and -cov

sriniv27 committed:
make format

sriniv27 committed:
remove unused imports

sriniv27 committed:
make format

sriniv27 committed:
Merge branch 'develop' of https://github.com/jrnl-org/jrnl into develop

sriniv27 committed:
rename --override to --config-override

sriniv27 committed:
move override implementation into own module

sriniv27 committed:
begin TDD of dot notated overrides

sriniv27 committed:
rewrite behavior scenario

sriniv27 committed:
implement recursive config overrides

sriniv27 committed:
clean up unittests

sriniv27 committed:
iterate on behave step

sriniv27 committed:
make format

sriniv27 committed:
cleanup

sriniv27 committed:
move override behave tests out of core

sriniv27 committed:
refactor recursive code

sriniv27 committed:
make format

sriniv27 committed:
code cleanup

sriniv27 committed:
remove unused import

sriniv27 committed:
update test config

sriniv27 committed:
rewrite test for better mock call expect

sriniv27 committed:
make format

sriniv27 committed:
binary search misbehaving windows test

sriniv27 committed:
Merge pull request #1 from sriniv27/separate-override-behave-test-from-core

separate-override-behave-test-from-core

sriniv27 committed:
unittest multiple overrides

sriniv27 committed:
uncomment dot notation unittest

sriniv27 committed:
add multiple override scenario spec

sriniv27 committed:
make format

sriniv27 committed:
make format

sriniv27 committed:
update unittests for new syntax

sriniv27 committed:
update integ tests for new syntax

sriniv27 committed:
update gitignore

sriniv27 committed:
guard override application

sriniv27 committed:
deserialize function as return type

sriniv27 committed:
make format

sriniv27 committed:
organize deserialization unittests

sriniv27 committed:
better, more specific behave tests

sriniv27 committed:
test different editor launch commands

sriniv27 committed:
formatting

sriniv27 committed:
handle datatypes in deserialization and update helptext

sriniv27 committed:
stick to config convention in testbed

sriniv27 committed:
update tests ith better verifications

sriniv27 committed:
update tests with better verifications

make format

space

review feedbac

sriniv27 committed:
update tests ith better verifications

sriniv27 committed:
make format

sriniv27 committed:
make format

sriniv27 committed:
space

sriniv27 committed:
space

sriniv27 committed:
review feedbac

sriniv27 committed:
Merge pull request #2 from sriniv27/double-delimited-text-input-for-config-override

Double delimited text input for config override

dependabot[bot] committed:
Bump pytest from 6.2.1 to 6.2.2

Bumps [pytest](https://github.com/pytest-dev/pytest) from 6.2.1 to 6.2.2.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/6.2.1...6.2.2)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump pytest from 6.2.1 to 6.2.2

Bumps [pytest](https://github.com/pytest-dev/pytest) from 6.2.1 to 6.2.2.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/6.2.1...6.2.2)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump keyring from 21.8.0 to 22.0.1

Bumps [keyring](https://github.com/jaraco/keyring) from 21.8.0 to 22.0.1.
- [Release notes](https://github.com/jaraco/keyring/releases)
- [Changelog](https://github.com/jaraco/keyring/blob/main/CHANGES.rst)
- [Commits](https://github.com/jaraco/keyring/compare/v21.8.0...v22.0.1)

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump keyring from 21.8.0 to 22.0.1

Bumps [keyring](https://github.com/jaraco/keyring) from 21.8.0 to 22.0.1.
- [Release notes](https://github.com/jaraco/keyring/releases)
- [Changelog](https://github.com/jaraco/keyring/blob/main/CHANGES.rst)
- [Commits](https://github.com/jaraco/keyring/compare/v21.8.0...v22.0.1)

Signed-off-by: dependabot[bot] <support@github.com>

sriniv27 committed:
make format

sriniv27 committed:
skip on win

sriniv27 committed:
skip on win

sriniv27 committed:
skip on win

sriniv27 committed:
update deps

sriniv27 committed:
update deps

sriniv27 committed:
Merge pull request #3 from sriniv27/dependabot/pip/pytest-6.2.2

Bump pytest from 6.2.1 to 6.2.2

sriniv27 committed:
Merge pull request #4 from sriniv27/dependabot/pip/keyring-22.0.1

Bump keyring from 21.8.0 to 22.0.1

sriniv27 committed:
refactor deserialization

organize test_parse_args

make format

sriniv27 committed:
refactor deserialization

organize test_parse_args

make format

sriniv27 committed:
Merge branch 'develop' of https://github.com/sriniv27/jrnl into develop

sriniv27 committed:
Merge branch 'develop' of https://github.com/sriniv27/jrnl into develop

sriniv27 committed:
make format

sriniv27 committed:
document apply_overrides

sriniv27 committed:
update gitignore

sriniv27 committed:
document config-override enhancement

wren reviewed (changes requested):
This is looking great!

I'd like to change the approach, so please see the ticket for my notes.

Thank you for putting so much work into this!

sriniv27 committed:
Merge branch 'develop' into lockfile-merge

sriniv27 committed:
Simplify config override syntax (#5)

* update tests and expected behavior

* clean up arg parsing tests

* update deserialization

* update deserialization

* config argparse action

* update override application logic

* update tests; delete unused imports

* override param must be list

* update docstring

* update test input to SUT

* update remaining override unittests

* make format

* forgot to update CLI syntax

sriniv27 committed:
update documentation to sphinx style

sriniv27 committed:
variable renames

dependabot[bot] committed:
Bump pytz from 2020.5 to 2021.1 (#6)

Bumps [pytz](https://github.com/stub42/pytz) from 2020.5 to 2021.1.
- [Release notes](https://github.com/stub42/pytz/releases)
- [Commits](https://github.com/stub42/pytz/compare/release_2020.5...release_2021.1)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sriniv27 committed:
Lockfile merge (#7)

* Add brew and gitter badges to README

* Update changelog [ci skip]

* Make journal selection behavior more consistent when there's a colon with no date (#1164)

* Bump pytest from 6.2.1 to 6.2.2 (#1167)

Bumps [pytest](https://github.com/pytest-dev/pytest) from 6.2.1 to 6.2.2.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/6.2.1...6.2.2)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump keyring from 21.8.0 to 22.0.1 (#1168)

Bumps [keyring](https://github.com/jaraco/keyring) from 21.8.0 to 22.0.1.
- [Release notes](https://github.com/jaraco/keyring/releases)
- [Changelog](https://github.com/jaraco/keyring/blob/main/CHANGES.rst)
- [Commits](https://github.com/jaraco/keyring/compare/v21.8.0...v22.0.1)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Update changelog [ci skip]

Co-authored-by: Micah Jerome Ellison <micah.jerome.ellison@gmail.com>
Co-authored-by: Jrnl Bot <jrnl.bot@gmail.com>
Co-authored-by: Jonathan Wren <jonathan@nowandwren.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sriniv27 committed:
Simplify config override syntax (#8)

* update tests and expected behavior

* clean up arg parsing tests

* update deserialization

* update deserialization

* config argparse action

* update override application logic

* update tests; delete unused imports

* override param must be list

* update docstring

* update test input to SUT

* update remaining override unittests

* make format

* forgot to update CLI syntax

sriniv27 committed:
Merge branch 'develop' into lockfile-merge

sriniv27 committed:
formatting

sriniv27 committed:
Update pyproject.toml

sriniv27 committed:
update lockfile to remove pytest-cov and pytest-mock deps

wren commented on docs/advanced.md:
```diff
@@ -61,6 +61,28 @@ and can be edited with a plain text editor.
 
+### Modifying Configurations from the Command line 
+
+You can override a configuration field for the current instance of `jrnl` using `--config-override CONFIG_KEY:CONFIG_VALUE` where `CONFIG_KEY` is a valid configuration field, specified in dot-notation and `CONFIG_VALUE` is the (valid) desired override value.
```
This still uses the colon format from before. Can you update for the new syntax?

micahellison commented on features/overrides.feature:
```diff
@@ -0,0 +1,53 @@
+Feature: Implementing Runtime Overrides for Select Configuration Keys
+
+        Scenario: Override configured editor with built-in input === editor:''
+        Given we use the config "tiny.yaml"
```
Could you use the basic_ configs for each journal type, like in tags.feature? Then remove tiny.yaml? It'd be nice to need keep the number of configs down.

micahellison commented on features/overrides.feature:
```diff
@@ -0,0 +1,53 @@
+        ┖─────────────────────┘
+        """
+
+        @skip_win
```
I removed this @skip_win on my Windows machine and this test is passing

micahellison commented on features/overrides.feature:
```diff
@@ -0,0 +1,53 @@
+        When we run jrnl with -1 --config-override colors.body blue
+        Then the runtime config should have colors.body set to blue
+
+        @skip_win 
```
I do see that this test and the test below are failing in Windows, but I think it would be good to ensure they are working. Let me know if you'd like some help troubleshooting this.

sriniv27 replied:
So I removed the tags and the tests fail in the CI runner with a the following UnicodeEncodeError: 

`UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f4af'` which is the 100 emoji. That's about as far as I've been able to get so far. 

Here's the stack trace I'm examining: 
```
 File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\runpy.py", line 193, in _run_module_as_main
FAILURE in step 'the runtime config should have editor set to nano':
  Feature:  Implementing Runtime Overrides for Select Configuration Keys
  Scenario: Apply multiple config overrides
Exception UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f4af' in position 533: character maps to <undefined>
    "__main__", mod_spec)
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "D:\a\jrnl\jrnl\.venv\Scripts\behave.exe\__main__.py", line 7, in <module>
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\__main__.py", line 183, in main
    return run_behave(config)
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\__main__.py", line 127, in run_behave
    failed = runner.run()
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\runner.py", line 804, in run
    return self.run_with_paths()
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\runner.py", line 824, in run_with_paths
    return self.run_model()
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\runner.py", line 626, in run_model
    failed = feature.run(self)
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\model.py", line 350, in run
    formatter.eof()
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\formatter\progress.py", line 86, in eof
    self.report_failures()
  File "d:\a\jrnl\jrnl\.venv\lib\site-packages\behave\formatter\progress.py", line 133, in report_failures
    self.stream.write(u"%s\n" % step.error_message)
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f4af' in position 533: character maps to <undefined>
Error: Process completed with exit code 1.
```

micahellison replied:
I created #1200 to troubleshoot, but it looks like all these passed in the CI this time. Could you try removing them and confirming?

wren commented on features/steps/override.py:
```diff
@@ -0,0 +1,123 @@
+    # fmt: on
+
+
+@then("the stdin prompt must be launched")
```
Can we change this wording to match the editor one? So, something like "the stdin prompt should have been called".

wren commented on features/steps/override.py:
```diff
@@ -0,0 +1,123 @@
+    return wrapper
+
+
+@given("we use the config {config_file}")
```
Please try to re-use the functions in `core.py` when possible. We shouldn't copy them into a new file if we can avoid it.

sriniv27 replied:
oh sure, thanks. I'll be more careful

wren commented on features/steps/override.py:
```diff
@@ -0,0 +1,123 @@
+def config_override(context, key_as_dots: str, override_value: str):
+    key_as_vec = key_as_dots.split(".")
+
+    def _mock_callback(**args):
```
Is this used anywhere?

sriniv27 replied:
good catch. my bad

wren commented on features/steps/override.py:
```diff
@@ -0,0 +1,123 @@
+
+    # fmt: off
+    try: 
+        with \
```
Can we reduce the amount of mocks we put into this test? If we mock too many things, we lose the value of these tests (since we could be mocking out bugs).

wren commented on features/steps/override.py:
```diff
@@ -0,0 +1,123 @@
+    context.cfg = cfg
+
+
+@then("the runtime config should have {key_as_dots} set to {override_value}")
```
This is a great idea for a test step!

Can it be merged with the other config step in `core.py`?

sriniv27 replied:
done!

micahellison commented on jrnl/override.py:
```diff
@@ -0,0 +1,47 @@
+# import logging
+def apply_overrides(overrides: list, base_config: dict) -> dict:
```
There is a similar `update_config` method in config.py that merges configuration of individual journals into the global config. Is there a reason to keep it separate?

sriniv27 replied:
I'd only ever seen it used in the context of applying the --encrypt and --decrypt postconfig commands, but I don't see any argument against just calling update_config after unpacking . Lemme work on updating the implementation.

wren commented on jrnl/jrnl.py:
```diff
@@ -49,6 +50,11 @@ def run(args):
+    # Apply config overrides
+    overrides = args.config_override
+    if overrides:
+        config = apply_overrides(overrides, config)
```
This should be farther up in the file (right after line 40 `config = scope_config(config, args.journal_name)`) since there are commands that require the config. If the overrides stay here, then none of those commands (import, encrypt, decrypt, etc) will work with overrides.

sriniv27 replied:
good to know, and thanks for catching that. Sounds like behavior I should have tested, so I can add some

micahellison commented on pyproject.toml:
```diff
@@ -38,15 +38,16 @@ pyyaml = ">=5.1"
 toml = ">=0.10"
 pyflakes = ">=2.2.0"
 pytest = ">=6.2"
+mock = "^4.0.3"
```
Is there a reason for this? `mock` is in the standard library.

sriniv27 replied:
removed and adjusted import declaration

micahellison commented on tests/test_config.py:
```diff
@@ -0,0 +1,97 @@
+    mock_load_or_install.return_value = minimal_config
+    mock_isatty.return_value = True
+
+    cli_args = shlex.split('--config-override editor "nano"')
```
I think we should we be using os_compat split_args instead of shlex.split to be cross-platform

sriniv27 replied:
good to know! For the purpose of this test, however, this is moot since I ended up deleting test_config.py altogether

micahellison commented on tests/test_config.py:
```diff
@@ -0,0 +1,97 @@
+    mock_isatty,
+    minimal_config,
+    expected_override,
+    capsys,
```
I think this `capsys` is an extra argument here and in the next function

micahellison commented on tests/test_config.py:
```diff
@@ -0,0 +1,97 @@
+@mock.patch("sys.stdin.isatty")
+@mock.patch.object(install, "load_or_install_jrnl")
+@mock.patch("subprocess.call")
+def test_override_configured_editor(
```
This looks more like a behavorial test than a unit test - I believe it's already covered in the feature file. Ideally, the pytest unit test should be much smaller and testing individual functions or "units" of code, rather than large chunks of the program.

micahellison commented on tests/test_config.py:
```diff
@@ -0,0 +1,97 @@
+    "jrnl.install.load_or_install_jrnl", wraps=jrnl.install.load_or_install_jrnl
+)
+@mock.patch("subprocess.call")
+def test_override_configured_colors(
```
This test has similar issues to the test above.

sriniv27 replied:
I just ended up deleting the unittest file. The behavior is covered in the feature files

wren commented on jrnl/override.py:
```diff
@@ -0,0 +1,47 @@
+    for pairs in overrides:
+
+        key_as_dots, override_value = list(pairs.items())[0]
+        keys = key_as_dots.split(".")
```
Can the dot notation resolve logic be a standalone function? This will make it easier to test.

sriniv27 replied:
can do

wren commented on jrnl/args.py:
```diff
@@ -17,6 +17,48 @@
+    cfg_value = input[1]
+    cfg_value = cfg_value.strip()
+
+    # Convert numbers and booleans
```
Instead of doing our own type coercion here, we should rely on the yaml library we already use. What if we don't choose a type until we know where the value is going? Then we can match the type to the destination (which has already been parsed by the yaml library).

sriniv27 replied:
I'm forwarding the input to `yaml.load` and returning that output. Let me know if that is not the direction you'd like to take it and we can have a discussion on implementing it

wren commented on jrnl/args.py:
```diff
@@ -17,6 +17,48 @@
+    elif cfg_value.lower() == "false":
+        cfg_value = False
+
+    runtime_modifications[cfg_key] = cfg_value
```
Does this handle keys with dots?

sriniv27 replied:
Yes, it stores the keys in dot notation. Those are unpacked into the config tree in apply_overrides

wren commented on jrnl/args.py:
```diff
@@ -17,6 +17,48 @@
+    return runtime_modifications
+
+
+class ConfigurationAction(argparse.Action):
```
Is this still needed?

sriniv27 replied:
It is. This is the action for --config-pverride argument. I intended to have an action for config modification operations, i.e. set, unset, and overrides.

wren replied:
I've been going back and forth on this, but I don't this deserialization should happen inside of arg parsing. At this point in the control flow, we don't have the config loaded yet, so we can't validate anything against it.

Can you move the function `deserialize_config_args` into config, and take out the arg parsing action?

wren reviewed (changes requested):
Getting there! Just a few more notes. Thank you!

micahellison reviewed (changes requested):
Thanks for all this work! I have a few questions and recommendations in my review.

sriniv27 committed:
update docs

sriniv27 committed:
reuse existing mock; delete unneeded code

sriniv27 committed:
move overrides earlier in the execution

use existing configs instead of custom

make format

clean up imports

sriniv27 committed:
update for passworded access

context.parser -> parsed_args

sriniv27 committed:
test that no editor is launched

sriniv27 committed:
remove unnecessary mocks

sriniv27 committed:
rename variable for intent

sriniv27 committed:
reinstate getpass deletion

sriniv27 committed:
update gitignore

sriniv27 committed:
capture failure mode

sriniv27 committed:
remove unneeded imports

sriniv27 committed:
renamed variable

sriniv27 committed:
delete redundant step

sriniv27 committed:
comment on step

sriniv27 committed:
clean up step behavior description

sriniv27 committed:
[WIP] lock down journal access behavior

sriniv27 committed:
skip -> wip

sriniv27 committed:
correct command for overriding journal via dot keys

sriniv27 committed:
update wip test for updating a "temp" journal and then reading baack its entries

sriniv27 committed:
remove "mock" from poetry file

sriniv27 committed:
make CI happy

sriniv27 committed:
complex behavior sequence for default journal override

sriniv27 committed:
separate out smaller pieces of logic

test that apply_overrides acts on base configuration and not the copy

sriniv27 committed:
defer modification of loaded configuration to update_config

remove unused fixtures

delete complicated UT since behavior is covered in overrides.feature integ test

delete redundant UT

sriniv27 committed:
Update .gitignore

sriniv27 committed:
remove skip_win

sriniv27 committed:
Merge branch 'lockfile-merge' of https://github.com/sriniv27/jrnl into lockfile-merge

sriniv27 committed:
forward override unpacking to yaml library

sriniv27 committed:
merge config override step with existing config_var step in core

delete config_override step

unify step description syntax

sriniv27 committed:
delete unused and redundant code

sriniv27 committed:
rebases are hard

sriniv27 committed:
remove wip tag from test

wren commented:
@sriniv27 Hi! I see that you requested review, but there are still commits happening, and activity in the related issue. Is this ready for review?

It's fine either way, I just didn't want to comment on things if you're still working on them.

sriniv27 commented:
I've only been making commits here to directly address review comments and smaller housekeeping things. I'd call this ready for review.

wren commented on features/overrides.feature:
```diff
@@ -0,0 +1,101 @@
+        """
+        ┎─────╮2013-06-09 15:39
+        ┃ My  ╘═══════════════╕
+        ┃ fir st  ent ry.     │
```
The `fancy` format looks like it has some issues. We need to fix this (not in this PR; just putting a note so we don't forget).

sriniv27 replied:
created #1207 so there's something to mark this resolved

wren commented on features/overrides.feature:
```diff
@@ -0,0 +1,101 @@
+        Given we use the config "basic_dayone.yaml"
+        And we use the password "test" if prompted
+        When we run "jrnl --config-override journals.temp features/journals/simple.journal temp Sep 06 1969: @say Ni"
+       Then we should get no error
```
Spacing issue.

wren commented on jrnl/args.py:
```diff
@@ -314,8 +349,32 @@ def parse_args(args=[]):
 
-    return parser.parse_intermixed_args(args)
+    parsed_args = parser.parse_intermixed_args(args)
+    return parsed_args
```
Is this change doing anything?

wren commented on pyproject.toml:
```diff
@@ -38,6 +38,7 @@ pyyaml = ">=5.1"
 pytz = ">=2020"          # https://pythonhosted.org/pytz/#issues-limitations
 tzlocal = ">2.0, <3.0"   # https://github.com/regebro/tzlocal/blob/master/CHANGES.txt
 
+
```
Can you remove this?

wren commented on pyproject.toml:
```diff
@@ -46,7 +47,6 @@ toml = ">=0.10"
 pyflakes = ">=2.2.0"
 pytest = ">=6.2"
 yq = ">=2.11"
-
```
Can you put this back?

micahellison commented on tests/test_override.py:
```diff
@@ -0,0 +1,79 @@
+        "editor": "vim",
+        "journals": {"default": "/tmp/journals/journal.jrnl"},
+    }
+    yield cfg
```
Could you return instead of yield here? I don't see any iterating on this.

sriniv27 replied:
The intent of `yield` here is to return the result of the fixture setup. In pytest fixtures, teardown code follows the `yield` call. At least, that's how I interpreted [the doc page](https://docs.pytest.org/en/reorganize-docs/yieldfixture.html). 

That said, there isn't any teardown code here. So a `return` should be good enough.

micahellison commented on tests/test_override.py:
```diff
@@ -0,0 +1,79 @@
+
+
+def test_override_dot_notation(minimal_config):
+    cfg = minimal_config.copy()
```
No need to .copy() this -- the fixture is refreshed every time a test runs.

micahellison commented on tests/test_override.py:
```diff
@@ -0,0 +1,79 @@
+        {"journals.burner": "/tmp/journals/burner.jrnl"},
+    ]  # as returned by parse_args, saved in parser.config_override
+
+    cfg = apply_overrides(overrides, minimal_config.copy())
```
Also no need to .copy() here

wren commented on tests/test_override.py:
```diff
@@ -0,0 +1,79 @@
+    assert _get_config_node(minimal_config, "display_format") == None
+
+
+from jrnl.override import _get_key_and_value_from_pair
```
Can you move this to the top of the file?

wren commented on tests/test_override.py:
```diff
@@ -0,0 +1,79 @@
+    assert v == "fgh"
+
+
+from jrnl.override import _convert_dots_to_list
```
Can yo move this to the top of the file?

wren commented on tests/test_override.py:
```diff
@@ -0,0 +1,79 @@
+from jrnl.override import _convert_dots_to_list
+
+
+class TestDotNotationToList:
```
🚀

micahellison reviewed (changes requested):
Looks really close to finished! Thanks again for your work on this.

wren reviewed (changes requested):
This is looking really close to ready!

Please add me to "reqeust review" when it's ready again. Thanks again for all the hard work.

sriniv27 committed:
remove skipped tests for windows

sriniv27 committed:
Address code review

yield -> return

remove needless copy

adjust spacing

re-inline args return

reset packaging info to e6c0a16342a31bc2fca1e2e865646c27bd43d0b4

revert package version for this PR

sriniv27 committed:
consolidate imports

sriniv27 committed:
Defer config_override unpacking to dict *after* base config is loaded
store cli overrides without unpacking just yet

move deserialize_config_args to config module

delete custom Action class for config operations

apply [k,v] -> {k, v} for each override

update test data

update import

sriniv27 committed:
rename deserialize_config_args to better express intent

make format

sriniv27 reviewed (commented):
I think that's all the comments addressed

wren commented:
Awesome! I'll take a look again tonight.

micahellison reviewed (approved):
💯

wren reviewed (approved):
Looks great! Thank you! 🚀
